Add support for X448 and Ed448

mattcaswell opened the pull request:
This adds EVP layer support for X448 and Ed448. TLS support will come later.
<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING

Other than that, provide a description above this comment if there isn't one already

If this fixes a github issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

richsalz commented on crypto/ec/ecx_meth.c:
```diff
@@ -129,18 +242,21 @@ static int ecx_pub_decode(EVP_PKEY *pkey, X509_PUBKEY *pubkey)
+    GETPUBKEY(pubkeya, a);
+    GETPUBKEY(pubkeyb, b);
+
+    return !CRYPTO_memcmp(pubkeya, pubkeyb, KEYLEN(a));
```
shouldn't we compare key type first?
And I think the "==0" is more clear than "!"

mattcaswell replied:
We are guaranteed to have keys of the same type here. That check is done for us in EVP_PKEY_cmp. I changed it to "== 0"

richsalz commented on crypto/ec/ecx_meth.c:
```diff
@@ -163,25 +279,33 @@ static int ecx_priv_decode(EVP_PKEY *pkey, const PKCS8_PRIV_KEY_INFO *p8)
+    }
+
+    GETPRIVKEY(privkey, pkey);
+
```
Remove this blank line, same reason as above.

mattcaswell replied:
As above, this got moved around a bit anyway as part of the pkey.ptr change.

richsalz commented on crypto/ec/ecx_meth.c:
```diff
@@ -346,10 +517,12 @@ static int ecd_item_verify(EVP_MD_CTX *ctx, const ASN1_ITEM *it, void *asn,
-    /* Sanity check: make sure it is ED25519 with absent parameters */
-    if (OBJ_obj2nid(obj) != NID_ED25519 || ptype != V_ASN1_UNDEF) {
+    nid = OBJ_obj2nid(obj);
+    /* Sanity check: make sure it is ED25519/ED448 with absent parameters */
```
I think you should put the comment before the two assignments.  Or put a blank line before the comment.

mattcaswell replied:
Done.

richsalz commented on crypto/ec/ecx_meth.c:
```diff
@@ -372,14 +545,36 @@ static int ecd_item_sign(EVP_MD_CTX *ctx, const ASN1_ITEM *it, void *asn,
+{
+    /* Set algorithms identifiers */
+    X509_ALGOR_set0(alg1, OBJ_nid2obj(NID_ED448), V_ASN1_UNDEF, NULL);
+    if (alg2)
```
Explicit NULL test.

mattcaswell replied:
Fixed.

richsalz commented on crypto/evp/pmeth_gn.c:
```diff
@@ -150,23 +150,53 @@ int EVP_PKEY_CTX_get_keygen_info(EVP_PKEY_CTX *ctx, int idx)
+    EVP_PKEY_CTX *ctx = NULL;
+    EVP_PKEY *pkey = NULL;
+
+    /* Note for compatibility reasons we allow keylen to be < 0 in some cases */
```
Put a comma or colon after the word Note

mattcaswell replied:
Done

richsalz commented on doc/man3/EVP_PKEY_CTX_ctrl.pod:
```diff
@@ -51,7 +65,7 @@ B<optype> is a mask indicating which operations the control can be applied to.
 B<p1> and B<p2>.
 
-For B<cmd> = B<EVP_PKEY_CTRL_SET_MAC_KEY>, B<p1> is the length of the MAC key,
+For B<cmd> = B<EVP_PKEY_CTRL_SET_PRIV_KEY>, B<p1> is the length of the MAC key,
```
"When cmd is EVP_PKEY..."  Not the semi-math expression.

mattcaswell replied:
Done

richsalz commented on test/recipes/30-test_evp_data/evppkey.txt:
```diff
@@ -748,6 +748,40 @@ Result = KEYOP_INIT_ERROR
 Function = EVP_PKEY_verify_init
 Reason = operation not supported for this keytype
 
+Title = X448 test vectors (from RFC7748 6.2)
```
Maybe 448/25519 should be in a separate file?

mattcaswell replied:
Hmmm...not sure about that. It seems to make sense to keep it together with all of the other EVP_PKEY tests?

mattcaswell commented on crypto/ec/ecx_meth.c:
```diff
@@ -22,24 +22,88 @@
+#define GETPRIVKEY(priv, pk) \
+    do { \
+        if (IS25519((pk)->ameth->pkey_id)) { \
+            (priv) = ((X25519_KEY *)((pk)->pkey.ptr))->privkey; \
```
Yes, good idea. Actually union members for RSA, DSA and EC keys already exist there, so this makes a lot of sense. Done.

mattcaswell commented on crypto/ec/ecx_meth.c:
```diff
@@ -51,69 +115,118 @@ static int ecx_key_op(EVP_PKEY *pkey, int id, const X509_ALGOR *palg,
-                xkey->privkey[0] &= 248;
-                xkey->privkey[31] &= 127;
-                xkey->privkey[31] |= 64;
+                (*privkey)[0] &= 248;
```
I could do, but they appear as decimal in the RFC so I think this is probably easier to understand? Looking at this again though made me realise we need to do something similar for X448 (done).

richsalz replied:
okey then.

mattcaswell commented on crypto/ec/ecx_meth.c:
```diff
@@ -51,69 +115,118 @@ static int ecx_key_op(EVP_PKEY *pkey, int id, const X509_ALGOR *palg,
 
-    penc = OPENSSL_memdup(xkey->pubkey, X25519_KEYLEN);
+    GETPUBKEY(pubkey, pkey);
+
```
This got moved around a bit anyway as part of the pkey.ptr change.

mattcaswell commented:
Updated commits pushed addressing comments so far.

mattcaswell commented:
@richsalz - do you have any more comments on this PR? Ideally I'd like to get this into next week's alpha.

kroeckx commented on crypto/evp/pmeth_gn.c:
```diff
@@ -150,23 +150,53 @@ int EVP_PKEY_CTX_get_keygen_info(EVP_PKEY_CTX *ctx, int idx)
+    if (EVP_PKEY_keygen_init(ctx) <= 0)
+        goto err;
+    if (priv) {
+        if (EVP_PKEY_CTX_set_priv_key(ctx, key, (int)keylen) <= 0)
```
keylen is already an int

mattcaswell replied:
Good catch. Fixed.

kroeckx commented on crypto/evp/pmeth_gn.c:
```diff
@@ -150,23 +150,53 @@ int EVP_PKEY_CTX_get_keygen_info(EVP_PKEY_CTX *ctx, int idx)
+            goto err;
+    }
+    if (EVP_PKEY_keygen(ctx, &pkey) <= 0)
+        goto err;
```
It's currently not clear to me why this is called. If we pass it a key, why do we generate a key? Shouldn't the operation be either KEY_OP_PUBLIC or KEY_OP_PRIVATE?

kroeckx replied:
So it seems you turned the KEY_OP_KEYGEN parameter in something that doesn't generate a key anymore if a public or private key is already set and acts like the function was called with KEY_OP_PUBLIC or KEY_OP_PRIVATE instead. This seems to be rather unexpected behavior change of EVP_PKEY_keygen() to me and might break existing applications.

mattcaswell replied:
A couple of important points here: Firstly I'm not breaking anything or changing any behaviour here. This is *existing* behaviour. This is exactly the way the existing "new mac key" stuff does things, and that is implemented across multiple algorithms (including in 1.1.0) so we can't really change this. All I have done here is created a synonym for creating "new private keys" because all the algorithms we have done this for so far are MACs - this is the first one where that isn't the case: but there's nothing special or specific about MACs - its just about setting raw private key data. I have extended the concept a little to bring in public keys too, but the underlying concept is the same.

Secondly, key generation typically consists of two steps: setting up the parameters for the key, and setting the private/public key data. Usually the key data is randomly generated but in the old (existing) "new mac key" way of doing things key generation effectively reduces down to setting up the parameters, and then setting the key data to what has been supplied. I'm just reusing that approach here. In a sense "keygen" is a bit of a misnomer - think of it more as "preparing a key for use".

kroeckx commented on crypto/evp/pmeth_gn.c:
```diff
@@ -150,23 +150,53 @@ int EVP_PKEY_CTX_get_keygen_info(EVP_PKEY_CTX *ctx, int idx)
+    EVP_PKEY_CTX *ctx = NULL;
+    EVP_PKEY *pkey = NULL;
+
+    /* Note: for compatibility reasons we allow keylen to be < 0 in some cases */
```
Could you please expand that? What compatibility reasons?

mattcaswell replied:
Basically the old EVP_PKEY_new_mac_key() has this weird behaviour where, for HMAC only, if you set a keylen of -1 then it assumes that the (binary!) key is NUL terminated! We use that capability in the dgst app. In refreshing my memory on why this was the case I realised I went too far in the dgst app renaming EVP_PKEY_new_mac_key() to EVP_PKEY_new_priv_key(). It should stay as EVP_PKEY_new_mac_key() because of this oddity (the new functions take a size_t for the length so this isn't possible with them). I expanded out the comment to explain this better, and fixed the dgst app.

kroeckx commented on crypto/ec/ecx_meth.c:
```diff
@@ -16,30 +16,88 @@
+
+#define MAX_KEYLEN  ED448_KEYLEN
+
+#define GETPRIVKEY(priv, pk) \
```
Any reason why you prefer a define instead of a static (ossl_inline) function?

mattcaswell replied:
No, not really. Probably an inline function would be more readable, so I made that change.

kroeckx commented on doc/man3/EVP_PKEY_CTX_ctrl.pod:
```diff
@@ -67,8 +81,23 @@ B<pkeyutl>, B<genpkey> and B<req> commands.
+
+The EVP_PKEY_CTX_set_priv_key() macro sets the raw private key data to be used
+when during a key generation operation. Typically users would call
+L<EVP_PKEY_new_priv_key(3)> instead of this macro. It can be used with any of
```
If you set the private (or public) key you're really not generating a (new) key, so that sentence doesn't make sense to me.

It's also not clear to me why we need this as a public function.

("when during" doesn't make sense.)

mattcaswell replied:
I reworded this to make this clearer (and to try and explain what is meant by key generation better). Note: these aren't functions but macros. EVP_PKEY_CTX_set_mac_key() is already a public macro and EVP_PKEY_set_priv_key() is just a synonym for it (because it now covers more than just mac keys). The underlying ctrls *have* to be public. I could make the EVP_PKEY_set_priv_key() and EVP_PKEY_set_pub_key() macros private, but given that the ctrls are public it seems weird not too.

kroeckx commented on doc/man3/EVP_PKEY_new.pod:
```diff
@@ -23,6 +35,25 @@ EVP_PKEY_up_ref() increments the reference count of B<key>.
+B<key> points to the raw private key data for this B<EVP_PKEY> which should be
+of length B<keylen>. The length should be appropriate for the type of the key.
+The public key data will be automatically derived from the given private key
+data.
```
The RETURNS sections needs to be updated.

mattcaswell replied:
Fixed. Also fixed HISTORY while I was at it.

mattcaswell commented:
New commits pushed addressing the latest comments. I also rebased due to conflicts with master.

mattcaswell commented:
In developing the related TLS PR I found a few bugs in this. I've added a couple of commits to address those bugs. I also spotted some documentation which need updating so I've added a commit for that.

Still looking for an approval on this...

kroeckx commented:
Can you explain why ecx_key_op() also takes a for instance a KEY_OP_PRIVATE, and when that should be used and when KEY_OP_KEYGEN?

mattcaswell commented:
> Can you explain why ecx_key_op() also takes a for instance a KEY_OP_PRIVATE, and when that should be used and when KEY_OP_KEYGEN?

KEY_OP_PRIVATE is used when we are filling in an EVP_PKEY * during ASN.1 decoding. It performs additional checks such as ensuring that there no X509 algorithm parameters associated with the key (which would be an error). Note: all of this is *existing* functionality which I haven't changed except to extend it to work with more than just 25519.

kaduk commented on crypto/ec/ecx_meth.c:
```diff
@@ -16,30 +16,88 @@
+        if (pk->pkey.x448 == NULL)
+            return;
+        *priv = pk->pkey.x448->privkey;
+    } else {
```
I'm not sure how comfortable I am with an unqualified 'else' here (and similarly, later).

mattcaswell replied:
It really shouldn't be anything other than ED448 here. I could put an assert in if it makes you feel more comfortable? But I'm not sure whether it is worth it.

kaduk replied:
I mean, yes, this is all internal and I trust you do have gotten things right.
And if we tried switching to a switch statement, we'd get our choice of compilers warning us for no 'default' label or for an unreachable default label (if we use an enum and use all values of it)...

mattcaswell commented:
@kroeckx, do you have any more comments on this, or can I take @richsalz's approval?

kroeckx commented:
I will try to review it today or tomorrow.

kroeckx commented:
I have a little bit of a problem with arguments that say this is how it's already the case, I just do the same. I also don't see the old code do that in the ecx_key_op(). It really either copies the private or public key from the parameter it's called with or generated a new random key. In the new code I also really don't see the difference if it gets called with KEY_OP_PUBLIC, palg set to NULL and having the key in p / plen, instead of storing it in some other place and then calling KEY_OP_KEYGEN. It only seems to complicate and duplicate code.

What I do see is that the old EVP_PKEY_new_mac_key() did call EVP_PKEY_CTX_set_mac_key() to store the key at some temporary place and then call EVP_PKEY_keygen() to copy it again, just like you do now, and I'm not sure it's a good example.

So the only thing I can really say is that things are inconsistent. And the only reason I can see why those 2 new functions where added is for the test suite.

kroeckx commented:
Also, please don't invent a new file format for storing private and public keys.

kroeckx commented on crypto/include/internal/evp_int.h:
```diff
@@ -385,6 +410,9 @@ struct evp_pkey_st {
         struct ec_key_st *ec;   /* ECC */
+        CURVE25519_KEY *curve25519;
+        X448_KEY *x448;
+        ED448_KEY *ed448;
```
Please consider making an ECX_KEY where you allocate both members instead.

mattcaswell replied:
Good idea. I have no idea why I didn't do this to start with. It simplifies the code considerably.

mattcaswell commented:
> Also, please don't invent a new file format for storing private and public keys.

I'm not sure I understand this? I've not invented any file format in this PR. I assume you are referring to the specifics of how X448/Ed448 are encoded/decoded for PKCS8 private keys, or into a SubjectPublicKeyInfo structure for public keys. Both of these things are defined here:

https://tools.ietf.org/html/draft-ietf-curdle-pkix-07

This is the spec that we were already using for X25519/Ed25519, and I've just extended this to include X448/Ed448.

mattcaswell commented:
> I have a little bit of a problem with arguments that say this is how it's already the case, I just do the same. I also don't see the old code do that in the ecx_key_op(). It really either copies the private or public key from the parameter it's called with or generated a new random key. In the new code I also really don't see the difference if it gets called with KEY_OP_PUBLIC, palg set to NULL and having the key in p / plen, instead of storing it in some other place and then calling KEY_OP_KEYGEN. It only seems to complicate and duplicate code.
>
> What I do see is that the old EVP_PKEY_new_mac_key() did call EVP_PKEY_CTX_set_mac_key() to store the key at some temporary place and then call EVP_PKEY_keygen() to copy it again, just like you do now, and I'm not sure it's a good example.
>
> So the only thing I can really say is that things are inconsistent. And the only reason I can see why those 2 new functions where added is for the test suite.

I strongly believe the 2 new functions are required to make the EVP API usable for X25519/Ed25519/X448/Ed448. I made precisely this argument recently in #5453 (which complains about the unusability of the current API). Basically if you want to use those algorithms for anything other than TLS then its incredibly hard without these functions. For example there is currently no way of using an arbitrary X25519 private key without reading it in from a PKCS8 structure. The only way of supplying an externally provided X25519 public key is either reading it from a SubjectPublicKeyInfo or using an undocumented TLS specific function (EVP_PKEY_set1_tls_encodedpoint).

However, it seems we have disagreement on the implementation of these functions. In the interests of moving this PR forward, I propose that I remove them from this PR. They are not needed for TLS usage, so that will not slow down the review of #5470. I will then raise a new PR to introduce these functions (possibly done slightly differently to take account of your comments).

mattcaswell commented:
I'm closing this in favour of #5481
